### ISSUE TITLE
Add scopes for function and variable references

### ISSUE THREAD

**Issue by aeschli** [Microsoft/TypeScript-TmLanguage#149]
_From @chmln on April 15, 2016 20:18_

- VSCode Version: 1.0.0-insider
- OS Version: Windows 10

Steps to Reproduce:

1. Open any js file with abundant variables and functions
2. Compare with any other popular editor e.g. sublime, atom, brackets, etc

VS Code is a fantastic editor. However, for an editor that accentuates support for JavaScript, the syntax highlighting is less than sub-par. 

To illustrate, here's VS Code and Sublime using the same color scheme:
![vsc_](https://cloud.githubusercontent.com/assets/11352152/14573672/6ef0d336-0324-11e6-8b8a-6953b87cb469.PNG)
![subl3](https://cloud.githubusercontent.com/assets/11352152/14573673/6ef1c106-0324-11e6-8a62-d2f3fdc460f4.PNG)

The color scheme isn't at fault since the default (dark+) theme is not much better
![vsc](https://cloud.githubusercontent.com/assets/11352152/14573691/88886d72-0324-11e6-9eaa-23a3acde6431.PNG)

Here's the code in screenshot:
```javascript
init = function () {

        navigationCurrentMonth.className = 'flatpickr-current-month';
        calendarContainer.className = 'flatpickr-calendar';
        calendar.className = "flatpickr-days";
        
        instanceConfig = instanceConfig || {};

        self.config = {};
        self.element = element;

        for (var config in self.defaultConfig)
            self.config[config] =
                instanceConfig[config] ||
                self.element.dataset&&self.element.dataset[config.toLowerCase()] ||
                self.element.getAttribute("data-"+config)||
                self.defaultConfig[config];

        
        self.input = (self.config.wrap) ? element.querySelector("[data-input]") : element;    

        if(self.config.defaultDate){
            self.config.defaultDate = uDate(self.config.defaultDate);
        }

        if ( self.input.value || self.config.defaultDate )
            self.selectedDateObj = uDate(self.config.defaultDate||self.input.value);

       self.config.minDate =
            uDate(self.config.minDate === "today" ? new Date() : self.config.minDate, true);

       self.config.maxDate = uDate(self.config.maxDate, true);

        self.jumpToDate(self.selectedDateObj||self.config.minDate||currentDate);

        wrap();        
        buildCalendar();
        bind();

        if (!self.config.noCalendar)
            updateValue();
    };
```

Significant and *absolutely necessary* enhancements can be brought in by highlighting:
* Variables
* Function names
* Methods

_Copied from original issue: Microsoft/vscode#5361_

**Comment by aeschli** [comment]
The build-in default themes do - by design- not highlight variable, method and function references.
The dark/light theme colorize however declarations.

@chmln Which custom theme do you use?

**Comment by aeschli** [comment]
_From @chmln on April 20, 2016 21:12_

> Which custom theme to you use?

*One Dark*

I have tried all the popular color themes, and not a single one has colored variables, methods, etc. 

Logically, that can be explained by two scenarios:
* It is impossible to color variables, methods, and function references
* All the color scheme authors are closet notepad fans

**Comment by aeschli** [comment]
_From @chmln on April 20, 2016 21:12_

![1](https://cloud.githubusercontent.com/assets/11352152/14690555/9bd18810-071a-11e6-92de-57084534303b.png)
![2](https://cloud.githubusercontent.com/assets/11352152/14690552/9bb22dd0-071a-11e6-85a3-60bd9c475373.png)
![3](https://cloud.githubusercontent.com/assets/11352152/14690551/9bb1f89c-071a-11e6-8b70-76169bf985fb.png)
![5](https://cloud.githubusercontent.com/assets/11352152/14690553/9bb3ea26-071a-11e6-8a9f-e93c5a48df2e.png)
![4](https://cloud.githubusercontent.com/assets/11352152/14690554/9bb4de22-071a-11e6-9b59-e58d6d7fc229.png)

**Comment by aeschli** [comment]
_From @Tyriar on April 20, 2016 21:51_

The root issue here is that https://github.com/Microsoft/TypeScript-TmLanguage does not scope variables and method calls.

**Comment by aeschli** [comment]
_From @chmln on April 20, 2016 22:2_

That explains it.
I see issues like https://github.com/Microsoft/TypeScript-TmLanguage/issues/72 so I surely can't be the first one to notice those shortcomings.

**Comment by aeschli** [comment]
Looking at Atom JS grammar it seems they use the `support.function` scope for function references.
For variable references it's `variables.other.object`.
We would still need a scope that stands for just the variable declaration.
Moving this to the TM/JS grammar team for consideration. I personally think we should stick to what we have.

**Comment by gaui** [comment]
:+1: I was shocked to see how poor JavaScript syntax highlighting is compared to Sublime and Atom. Hope this gets taken care of.

**Comment by adambuczynski** [comment]
It was indeed rather shocking to see the sub-par syntax highlighting for typescript when compared to javascript.

I might add that syntax highlighting in Atom for TypeScript files is bad too, as the `atom-typescript` package uses the grammar files from this repository. As such, comparing `.ts` code to `.js` code results in something like this (ignore the color differences, just see how much more detail there is):

TS:
![image](https://cloud.githubusercontent.com/assets/490562/16287333/6ff88718-3936-11e6-9069-29b9b33ba073.png)

JS:
![image](https://cloud.githubusercontent.com/assets/490562/16287340/77809fd4-3936-11e6-8ae4-a5cd1f670fd7.png)

Over the past few months, syntax highlighting in the `language-javascript` package has improved significantly, with good support for ES6 features and much improved highlighting of variables, properties, methods calls, etc.

Maybe it's possible to convert/copy some of the improved grammar rules from JS and use them for TS as well?

**Comment by whitecolor** [comment]
+1

**Comment by glen-84** [comment]
@whitecolor Please don't post `+1` comments. Use reactions instead.

**Comment by sheetalkamat** [comment]
Fixed with #270
